### DISCUSSION TITLE
Something is acting weird when I execute my stored procedure

### DISCUSSION THREAD

**Discussion by MANLW-Vestas** [sqlalchemy/sqlalchemy#8407] [Usage Questions] [answered]
I'm working with SqlAlchemy to access my Microsoft Database. I have made a stored procedure that does some relevant inserts and selects some other things. I have tested some other SP with parameters, and they worked as intended.

My problem is only present when I try to execute that specific SP through my python script. The SP has one parameter, that is a User-Defined Table Type. The SP then takes that table input and processes that accordingly, and returns the specified elements.
The thing that confuses me the most, is the fact that not errors are being thrown when I execute it in the script, yet nothing is happening. If I then print out the exact same command that the script is trying to execute, and copies that over to SSMS, then it works without any problems, and is doing exactly what it's supposed to.

The SQL-side:

```SQL
CREATE PROCEDURE dbo.StoredProcedure
     @param1 dbo.User-Defined Table Type READONLY
AS
SET NOCOUNT ON;
     INSERT INTO Table1 (Col1, Col2) SELECT TOP 1 Col1, Col2 FROM @param1

     DECLARE @TempVar INT = (SELECT TOP 1 Id FROM Table1 ORDER BY Id DESC)
     
     INSERT INTO Table2
          (Col1, Col2, Col3)
          SELECT DISCTICT p.Col1, p.Col2, p.Col3
          FROM @param1 as p
          LEFT JOIN Table2 t2 ON t2.Col1 = p.Col1
               WHERE t2.Col1 IS NULL

     INSERT INTO Table3 (Col1, Col2, Col3, Col4)
          SELECT @TempVar, Col1, Col2, Col3, Col4 FORM @Param1

     SELECT DISTINCT Id From @param1
GO
```

My Python script code:
```Python
engine = create_engine(conn_str)

engine.execute("""
     DECLARE @Params User-Defined Table Type

     INSERT INTO @Params (col1, col2, col3, col4)
     VALUES (val1, val2, val3, val4), (val1, val2, val3, val4), (val1, val2, val3, val4)

     EXECUTE _SP_ @Param1=@Params
""")
```

This specific setup does not work correct in the script. It runs as if it have done it successfully, yet nothing has been uploaded. I've noticed that one of the tables that have an id that uses AUTO INCREMENT is actually incrementing even though no row was added. So if I run the exact same query, but in SSMS, I can see a gap in the ID's.

I have run out of ideas on what is causing this, so I hope that someone here is able to give me an answer.

UPDATE:
The full query (I.e a lot more insert statements in full query) in Python takes around 5 seconds to execute, while if i copy it to SSMS it takes around 30 seconds

**Comment by CaselIT** [comment]
Hi,

Have you tried commiting the connection or settings autocommit mode as the engine isolation level?

**Reply by MANLW-Vestas** [reply]
I have not tried any things like that, i’ll try some different settings and see if that makes any difference

**Comment by lelit** [comment]
In your Python script, nothing is executed on the engine, you just redeclare the variable `engine`...

**Reply by MANLW-Vestas** [reply]
My bad, forgot to put the .execute in the post. Thanks for noticing. I have updated the post

**Reply by CaselIT** [reply]
Note that engine.execute is deprecated, and it should raise deprecantion warnings

**Reply by CaselIT** [reply]
Same with passing a string to execute. You should use `text()` wirh execute, or `connection.exec_driver_sql()` to directly pass a string to the dbapi

**Comment by zzzeek** [comment]
hi  -

You may also need to use [cursor.callproc()](https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=callproc#calling-stored-procedures-and-user-defined-functions) directly for special statements like this.   at that point you can get help from pyodbc people if there are still problems.

**Comment by gordthompson** [comment, marked as answer]
> I've noticed that one of the tables that have an id that uses AUTO INCREMENT is actually incrementing even though no row was added.

So one or more inserts have been executed but not committed.

Your stored procedure may be throwing an error that you are not seeing because your anonymous code block does not start with `SET NOCOUNT ON;`. Try this:

```python
stmt = """\
SET NOCOUNT ON;
DECLARE @Params User-Defined Table Type

INSERT INTO @Params (col1, col2, col3, col4)
VALUES (val1, val2, val3, val4), (val1, val2, val3, val4), (val1, val2, val3, val4)

EXECUTE _SP_ @Param1=@Params
"""

with engine.begin() as conn:
    result = conn.exec_driver_sql(stmt).all()
    print(result)
```

Also, re: "You may also need to use cursor.callproc() directly" - `callproc()` is an optional method in DBAPI 2.0 and is not implemented in pyodbc. Further details here:

https://github.com/mkleehammer/pyodbc/wiki/Calling-Stored-Procedures

**Reply by MANLW-Vestas** [reply]
Thanks, this worked! Using the .begin() and .exec_driver_sql() fixed.
I still wonder why .exec_driver_sql() wasn't registed by the intellisense
